{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://cropgenebank.sgrp.cgiar.org/images/file/learning_space/molecular_markers/volume2/04_Measures.pdf\n",
    "# Page 18?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#splitting it into parts:\n",
    "\n",
    "#calcualting h0\n",
    "def part_one(s, pop_1_A1A2, pop_2_A1A2, pop_3_A1A2=0, pop_4_A1A2=0, pop_5_A1A2=0):\n",
    "    Average = (pop_1_A1A2 + pop_2_A1A2 + pop_3_A1A2 + pop_4_A1A2+ pop_5_A1A2) /100\n",
    "    h_o = 1/s *(Average)\n",
    "    return h_o\n",
    "\n",
    "A = part_one(s = 3, pop_1_A1A2 = 30, pop_2_A1A2 = 20, pop_3_A1A2=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating n\n",
    "def part_two(s, pop_1_A1A1, pop_1_A1A2, pop_1_A2A2, pop_2_A1A1, pop_2_A1A2, pop_2_A2A2,\n",
    "             pop_3_A1A1 = 0, pop_3_A1A2 = 0, pop_3_A2A2 = 0,\n",
    "             pop_4_A1A1 = 0, pop_4_A1A2 = 0, pop_4_A2A2 = 0,\n",
    "             pop_5_A1A1 = 0, pop_5_A1A2 = 0, pop_5_A2A2 = 0):\n",
    "    try: \n",
    "        pop_1_a = (pop_1_A1A1 + pop_1_A1A2 + pop_1_A2A2)/3\n",
    "        pop_2_a = (pop_2_A1A1 + pop_2_A1A2 + pop_2_A2A2)/3\n",
    "        pop_3_a = (pop_3_A1A1 + pop_3_A1A2 + pop_3_A2A2)/3\n",
    "        pop_4_a = (pop_4_A1A1 + pop_4_A1A2 + pop_4_A2A2)/3\n",
    "        pop_5_a = (pop_5_A1A1 + pop_5_A1A2 + pop_5_A2A2)/3\n",
    "        \n",
    "    except:\n",
    "        pop_3_a = 0\n",
    "        pop_4_a = 0\n",
    "        pop_5_a = 0\n",
    "        \n",
    "    n = (pop_1_a + pop_2_a + pop_3_a + pop_4_a + pop_5_a)/s\n",
    "    return n\n",
    "A =part_two(s = 3, pop_1_A1A1 = 20, pop_1_A1A2 =30, pop_1_A2A2 =50, pop_2_A1A1 = 10, pop_2_A1A2 =20, pop_2_A2A2 = 70, pop_3_A1A1 = 60, pop_3_A1A2 = 10, pop_3_A2A2 = 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating p_q\n",
    "def part_three(pop_1_A1A1, pop_1_A2A2, pop_2_A1A1, pop_2_A2A2, pop_3_A1A1=0, pop_3_A2A2 =0,\n",
    "               pop_4_A1A1 = 0, pop_4_A2A2 =0, pop_5_A1A1=0, pop_5_A2A2=0):\n",
    "    pq_1 = ((pop_1_A1A1 **2) + (pop_1_A2A2 **2))\n",
    "    pq_2 = ((pop_2_A1A1 **2) + (pop_2_A2A2 **2))\n",
    "    pq_3 = ((pop_3_A1A1 **2) + (pop_3_A2A2 **2))\n",
    "    pq_4 = ((pop_4_A1A1 **2) + (pop_4_A2A2 **2))\n",
    "    pq_5 = ((pop_5_A1A1 **2) + (pop_5_A2A2 **2))\n",
    "\n",
    "    total_pq = pq_1+ pq_2 + pq_3 +pq_4+ pq_5\n",
    "    return total_pq\n",
    "\n",
    "a = part_three(pop_1_A1A1 = 0.35, pop_1_A2A2 = 0.65, pop_2_A1A1 =0.2, pop_2_A2A2 = 0.8,\n",
    "               pop_3_A1A1=0.65, pop_3_A2A2 =0.35,)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#caluating hs\n",
    "#calulating population diversity\n",
    "def part_four(n, s, total_pq, ho):\n",
    "    A = n/(n-1)\n",
    "    B = (1-((1/s) * (total_pq)))\n",
    "    C = (ho/(2 * n))\n",
    "    hs = A * (B - C)\n",
    "    return hs\n",
    "\n",
    "a = part_four(n =33.33, s =3, total_pq =1.77, ho = 0.199)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def part_five(s,pop_1_p, pop_1_q, pop_2_p, pop_2_q, pop_3_p=0, pop_3_q =0,\n",
    "               pop_4_p = 0, pop_4_q =0, pop_5_p=0, pop_5_q=0):\n",
    "    \n",
    "    T = ((1/s *(pop_1_p))**2 + (1/s *(pop_1_q))**2 + (1/s *(pop_2_p))**2 + (1/s *(pop_2_q))**2 \n",
    "    + (1/s *(pop_3_p))**2 + (1/s *(pop_3_q))**2 + (1/s *(pop_4_p))**2 + (1/s *(pop_4_q))**2 \n",
    "    + (1/s *(pop_5_p))**2 + (1/s *(pop_5_q))**2)\n",
    "    return T\n",
    "\n",
    "a = part_five(s = 3,pop_1_p = 0.35, pop_1_q = 0.65, pop_2_p = 0.2, pop_2_q =0.8, pop_3_p=0.65, pop_3_q =0.35)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating ht\n",
    "#calculating total diversity\n",
    "\n",
    "def part_six(T, hs, n, s, ho):\n",
    "    ht = (1-T) + (hs/(n*s)) - (ho/(2*n*s))\n",
    "    return ht\n",
    "\n",
    "a = part_six(T= 0.1966, hs = 0.4196, n =33.33 , s = 3, ho =0.1999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcualting gst\n",
    "#measure of differentiation in terms of\n",
    "#alleles per locus in two or more populations\n",
    "def part_seven(hs, ht):\n",
    "    gst = 1 - (hs/ht)\n",
    "    return gst\n",
    "\n",
    "a = part_seven(hs = 0.41960, ht = 0.80659)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Final genetic diversity formula\n",
    "## Measure of differentiation in terms of alleles per locus in two or more population\n",
    "## Ranges from 0 to 1. A negative value may be obtained if an error was made \n",
    "#   for sampling or an inappropriate marker system was used. \n",
    "## It can be used with codominant markers and restrictedly with dominant markers.\n",
    "## This is because it is a measure of heterozygosity. \n",
    "## To have a fair estimate of the real value, several generations are needed\n",
    "\n",
    "\n",
    "\n",
    "def genetic_diversity(s, pop_1_A1A1, pop_1_A1A2, pop_1_A2A2,\n",
    "                      pop_2_A1A1, pop_2_A1A2, pop_2_A2A2, \n",
    "                      pop_1_p, pop_1_q, pop_2_p, pop_2_q,\n",
    "                      pop_3_A1A1 = 0, pop_3_A1A2=0, pop_3_A2A2 = 0,\n",
    "                      pop_4_A1A1 = 0, pop_4_A1A2=0, pop_4_A2A2 = 0,\n",
    "                      pop_5_A1A1 = 0, pop_5_A1A2=0, pop_5_A2A2 = 0,\n",
    "                      pop_3_p=0, pop_3_q =0,pop_4_p = 0, pop_4_q =0, \n",
    "                      pop_5_p=0, pop_5_q=0):\n",
    "    Average = (pop_1_A1A2 + pop_2_A1A2 + pop_3_A1A2 + pop_4_A1A2+ pop_5_A1A2) /100\n",
    "    h_o = 1/s *(Average) \n",
    "\n",
    "    try: \n",
    "        pop_1_a = (pop_1_A1A1 + pop_1_A1A2 + pop_1_A2A2)/3\n",
    "        pop_2_a = (pop_2_A1A1 + pop_2_A1A2 + pop_2_A2A2)/3\n",
    "        pop_3_a = (pop_3_A1A1 + pop_3_A1A2 + pop_3_A2A2)/3\n",
    "        pop_4_a = (pop_4_A1A1 + pop_4_A1A2 + pop_4_A2A2)/3\n",
    "        pop_5_a = (pop_5_A1A1 + pop_5_A1A2 + pop_5_A2A2)/3\n",
    "    except:\n",
    "        pop_3_a = 0\n",
    "        pop_4_a = 0\n",
    "        pop_5_a = 0\n",
    "        \n",
    "    n = (pop_1_a + pop_2_a + pop_3_a + pop_4_a + pop_5_a)/s\n",
    "    \n",
    "  \n",
    "    pq_1 = ((pop_1_p **2) + (pop_1_q **2))\n",
    "    pq_2 = ((pop_2_p **2) + (pop_2_q **2))\n",
    "    pq_3 = ((pop_3_p **2) + (pop_3_q **2))\n",
    "    pq_4 = ((pop_4_p **2) + (pop_4_q **2))\n",
    "    pq_5 = ((pop_5_p **2) + (pop_5_q **2))\n",
    "    \n",
    "    total_pq = pq_1+ pq_2 + pq_3 +pq_4+ pq_5\n",
    "    \n",
    "    A = n/(n-1)\n",
    "    B = (1-((1/s) * (total_pq)))\n",
    "    C = (h_o/(2 * n))\n",
    "    hs = A * (B - C)\n",
    "    \n",
    "    T = ((1/s *(pop_1_p))**2 + (1/s *(pop_1_q))**2 + (1/s *(pop_2_p))**2 + (1/s *(pop_2_q))**2 \n",
    "         + (1/s *(pop_3_p))**2 + (1/s *(pop_3_q))**2 + (1/s *(pop_4_p))**2 + (1/s *(pop_4_q))**2 \n",
    "         + (1/s *(pop_5_p))**2 + (1/s *(pop_5_q))**2)\n",
    "    \n",
    "    ht = (1-T) + (hs/(n*s)) - (h_o/(2*n*s))\n",
    "    gst = 1 - (hs/ht)\n",
    "    return gst\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S = genetic_diversity(s=3, pop_1_A1A1=20, pop_1_A1A2=30, pop_1_A2A2=50,\n",
    "                      pop_2_A1A1=10, pop_2_A1A2=20, pop_2_A2A2=70, \n",
    "                      pop_1_p=0.35, pop_1_q=0.65, pop_2_p=0.20, pop_2_q=0.80,\n",
    "                      pop_3_A1A1 = 60, pop_3_A1A2=10, pop_3_A2A2 = 30, pop_3_p=0.65, pop_3_q =0.35)\n",
    "print(S)\n",
    "                      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#http://www.uwyo.edu/dbmcd/molmark/practica/fst.html\n",
    "\n",
    "#calculating F stats\n",
    "# in parts\n",
    "#freq of allele\n",
    "\n",
    "#Gene frequencies: \n",
    "#Each homozygote will have two alleles, each heterozygote will have one allele.\n",
    "def part_one(population_1, population_2, \n",
    "             pop1_AA, pop2_AA,\n",
    "             pop1_Aa, pop2_Aa,\n",
    "             pop1_aa, pop2_aa,\n",
    "             population_3 = 0, population_4 = 0, population_5 = 0, \n",
    "             pop3_AA = 0, pop4_AA = 0, pop5_AA = 0,\n",
    "             pop3_Aa = 0, pop4_Aa = 0, pop5_Aa = 0,\n",
    "             pop3_aa = 0, pop4_aa = 0, pop5_aa = 0):\n",
    "    p_1 = ((2 *pop1_AA ) + pop1_Aa) / (2 * population_1)\n",
    "    q_1= 1 - p_1\n",
    "    \n",
    "    p_2 = ((2* pop2_AA ) + pop2_Aa) / (2 * population_2)\n",
    "    q_2= 1 - p_2\n",
    "    \n",
    "    if population_3 != 0:\n",
    "        p_3 = ((2* pop3_AA ) + pop3_Aa) / (2 * population_3)\n",
    "        q_3 = 1 - p_3\n",
    "    else:\n",
    "        q_3 = 'N/A'\n",
    "        p_3 = 'N/A'\n",
    "        \n",
    "    if population_4 != 0:       \n",
    "        p_4 = ((2* pop4_AA ) + pop4_Aa) / (2 * population_4)\n",
    "        q_4 = 1 - p_4\n",
    "    else:\n",
    "        q_4 = 'N/A'\n",
    "        p_4 = 'N/A'\n",
    "        \n",
    "    if population_5 != 0: \n",
    "        p_5 = ((2* pop5_AA ) + pop5_Aa) / (2 * population_5)\n",
    "        q_5 = 1 - p_5\n",
    "    else:\n",
    "        q_5 = 'N/A'\n",
    "        p_5 = 'N/A'\n",
    "        \n",
    "\n",
    "        \n",
    "    return  p_1, q_1, p_2, q_2, p_3, q_3, p_4, q_4, p_5, q_5 \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0.5, 0.5, 0.65, 0.35, 0.35, 0.65, 'N/A', 'N/A', 'N/A', 'N/A')\n"
     ]
    }
   ],
   "source": [
    "A = part_one(population_1 = 500, population_2 = 100, \n",
    "             pop1_AA =125, pop2_AA =50,\n",
    "             pop1_Aa =250, pop2_Aa =30,\n",
    "             pop1_aa = 125, pop2_aa =20,\n",
    "             population_3 = 1000, \n",
    "             pop3_AA = 100,\n",
    "             pop3_Aa = 500,\n",
    "             pop3_aa = 400)\n",
    "print(A)\n",
    "             \n",
    "             "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Observed heterozygosity (here we are counting heterozygous genotypes\n",
    "def part_two(population_1, population_2,\n",
    "             pop1_Aa, pop2_Aa, \n",
    "             pop3_Aa = 0, pop4_Aa = 0, pop5_Aa = 0,\n",
    "             population_3 = 0, population_4 = 0, population_5 = 0):\n",
    "    \n",
    "    H_obs1 = pop1_Aa / population_1\n",
    "    H_obs2 = pop2_Aa / population_2\n",
    "    \n",
    "    if population_3 != 0:\n",
    "        H_obs3 = pop3_Aa / population_3\n",
    "    else:\n",
    "        H_obs3 = 'N/A'\n",
    "    \n",
    "    if population_4 != 0:\n",
    "        H_obs4 = pop4_Aa / population_4\n",
    "    else:\n",
    "        H_obs4 = 'N/A'\n",
    "    \n",
    "    if population_5 != 0:\n",
    "        H_obs5 = pop5_Aa / population_5\n",
    "    else:\n",
    "        H_obs5 = 'N/A'\n",
    "        \n",
    "    return H_obs1, H_obs2, H_obs3, H_obs4, H_obs5  \n",
    "    \n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0.5, 0.3, 0.5, 'N/A', 'N/A')\n"
     ]
    }
   ],
   "source": [
    "A = part_two(population_1 = 500, population_2 = 100,\n",
    "             pop1_Aa = 250, pop2_Aa = 30, \n",
    "             pop3_Aa = 500, population_3 = 1000)\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Expected heterozygosity (2pq)\n",
    "# 2 * p * q\n",
    "def part_three(p_1, q_1, p_2, q_2, \n",
    "               p_3 = 0, q_3 = 0, p_4 = 0, q_4 = 0, p_5 = 0, q_5 = 0 ):\n",
    "    \n",
    "    H_exp_1 = 2 * p_1 * q_1\n",
    "    H_exp_2 = 2 * p_2 * q_2\n",
    "    H_exp_3 = 2 * p_3 * q_3\n",
    "    H_exp_4 = 2 * p_4 * q_4\n",
    "    H_exp_5 = 2 * p_5 * q_5\n",
    "    \n",
    "    return H_exp_1, H_exp_2, H_exp_3, H_exp_4, H_exp_5\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0.5, 0.45499999999999996, 0.45499999999999996, 0, 0)\n"
     ]
    }
   ],
   "source": [
    "A = part_three(p_1 = 0.5, q_1 =0.5, p_2 = 0.65, q_2 = 0.35, \n",
    "               p_3 = 0.35, q_3 = 0.65)\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Population inbreeding coefficient [F = (Hexp - Hobs) /Hexp]\n",
    "def part_four(H_obs1, H_obs2, H_exp_1, H_exp_2,\n",
    "              H_obs3 = 0, H_obs4 = 0, H_obs5 = 0,\n",
    "             H_exp_3 = 0, H_exp_4 = 0, H_exp_5 = 0):\n",
    "    F_1 = (H_exp_1 - H_obs1 ) /H_exp_1\n",
    "    F_2 = (H_exp_2 - H_obs2 ) /H_exp_2\n",
    "    \n",
    "    if H_obs3 != 0:\n",
    "        F_3 = (H_exp_3 - H_obs3 ) /H_exp_3\n",
    "    else:\n",
    "        F_3 = 0\n",
    "    \n",
    "    if H_obs4 != 0:\n",
    "        F_4 = (H_exp_4 - H_obs4 ) /H_exp_4\n",
    "    else:\n",
    "        F_4 = 0\n",
    "    \n",
    "    if H_obs5 != 0:\n",
    "        F_5 = (H_exp_5 - H_obs5 ) /H_exp_5\n",
    "    else:\n",
    "        F_5 = 0\n",
    "    \n",
    "    return F_1, F_2,  F_3, F_4, F_5\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0.0, 0.3406448493373481, -0.09890134041787699, 0, 0)\n"
     ]
    }
   ],
   "source": [
    "A = part_four(H_obs1 = 0.5, H_obs2 = 0.3, H_exp_1 = 0.5, H_exp_2 = 0.45499, H_obs3 = 0.5, H_exp_3 = 0.4549999)\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#frequency of allele A over all populations\n",
    "def p_bar(p_1, p_2, population_1, population_2, N, p_3 = 0, p_4 = 0, p_5 = 0, population_3 = 0, population_4 = 0, population_5 = 0):\n",
    "    \n",
    "    if N == 5:\n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3) + p_4 * (2 * population_4)\n",
    "        +(p_5 * (2 * population_5))\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4 + population_5))\n",
    "    \n",
    "    elif N == 4:\n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3) + p_4 * (2 * population_4)\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4))\n",
    "        \n",
    "    if N == 3:\n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3)\n",
    "        p_bar = p_bar_1/  (2 *(population_1 + population_2 + population_3))\n",
    "        \n",
    "    else:\n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2)\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2))\n",
    "        \n",
    "    return p_bar\n",
    "        \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.415625\n"
     ]
    }
   ],
   "source": [
    "A = p_bar(p_1 = 0.5, p_2 = 0.65, p_3 = 0.35,\n",
    "          population_1 = 500, population_2 = 100, population_3= 1000 , N = 3)\n",
    "print(A)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#(frequency of allele a) over all populations\n",
    "def q_bar(N, pop1_aa, pop1_Aa, pop2_aa, pop2_Aa, population_1, population_2, \n",
    "          pop3_aa = 0, pop3_Aa = 0, pop4_aa = 0, pop4_Aa = 0, pop5_aa = 0, pop5_Aa = 0,\n",
    "         population_3 = 0, population_4 = 0, population_5 = 0):\n",
    "    \n",
    "    q_bar_1 = (2 *pop1_aa ) + pop1_Aa + (2 *pop2_aa ) + pop2_Aa + (2 *pop3_aa ) + pop3_Aa +  (2 *pop4_aa ) + pop4_Aa \n",
    "    +(2 *pop5_aa ) + pop5_Aa\n",
    "    q_bar = q_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4 + population_5))\n",
    "    \n",
    "    return q_bar\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.584375\n"
     ]
    }
   ],
   "source": [
    "A = q_bar(N = 3, pop1_aa = 125, pop1_Aa = 250, pop2_aa = 20, pop2_Aa = 30, population_1 = 500, population_2 = 100, \n",
    "          pop3_aa = 400, pop3_Aa = 500,\n",
    "         population_3 = 1000)\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Heterozygosity indices \n",
    "#(over individuals, subpopulations and total population)\n",
    "#HI based on observed heterozygosities in populations\n",
    "def H():\n",
    "    HI_i = (H_obs1 * population_1) + (H_obs2 * population_2) + (H_obs3 * population_3 ) + (H_obs4 * population_4 ) \n",
    "    + (H_obs5 * population_5 )\n",
    "    HI = HI_i / (population_1 + population_2 + population_3 + population_4 + population_5)\n",
    "    \n",
    "    HS_s = (H_exp1 * population_1) + (H_exp2 * population_2) + (H_exp3 * population_3 ) + (H_exp4 * population_4 ) \n",
    "    + (H_exp5 * population_5 )\n",
    "    HS = HS_s/ (population_1 + population_2 + population_3 + population_4 + population_5)\n",
    "    \n",
    "    HT = 2 * p_bar * q_bar\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###THE F-STATISTICS\n",
    "def F():\n",
    "    F_IS = (HS - HI) / HS\n",
    "    F_ST = (HT - HS)/ HT\n",
    "    F_IT = (HT -HI)/ HT\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "####FINAL fST eqn\n",
    "#put everything together\n",
    "\n",
    "#N = number of populations\n",
    "\n",
    "def Fst(N,population_1, population_2, \n",
    "             pop1_AA, pop2_AA,\n",
    "             pop1_Aa, pop2_Aa,\n",
    "             pop1_aa, pop2_aa,\n",
    "             population_3 = 0, population_4 = 0, population_5 = 0, \n",
    "             pop3_AA = 0, pop4_AA = 0, pop5_AA = 0,\n",
    "             pop3_Aa = 0, pop4_Aa = 0, pop5_Aa = 0,\n",
    "             pop3_aa = 0, pop4_aa = 0, pop5_aa = 0):\n",
    "    \n",
    "    if N == 2:\n",
    "        p_1 = ((2 *pop1_AA ) + pop1_Aa) / (2 * population_1)\n",
    "        q_1= 1 - p_1\n",
    "    \n",
    "        p_2 = ((2* pop2_AA ) + pop2_Aa) / (2 * population_2)\n",
    "        q_2= 1 - p_2\n",
    "        \n",
    "        H_obs1 = pop1_Aa / population_1\n",
    "        H_obs2 = pop2_Aa / population_2\n",
    "        \n",
    "        H_exp_1 = 2 * p_1 * q_1\n",
    "        H_exp_2 = 2 * p_2 * q_2\n",
    "        \n",
    "        F_1 = (H_exp_1 - H_obs1 ) /H_exp_1\n",
    "        F_2 = (H_exp_2 - H_obs2 ) /H_exp_2\n",
    "        \n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2)\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2))\n",
    "        \n",
    "        q_bar_1 = (2 *pop1_aa ) + pop1_Aa + (2 *pop2_aa ) + pop2_Aa  \n",
    "        q_bar = q_bar_1/ (2 *(population_1 + population_2))\n",
    "        \n",
    "        HI_i = (H_obs1 * population_1) + (H_obs2 * population_2)\n",
    "        HI = HI_i / (population_1 + population_2)\n",
    "    \n",
    "        HS_s = (H_exp_1 * population_1) + (H_exp_2 * population_2)\n",
    "        HS = HS_s/ (population_1 + population_2)\n",
    "        \n",
    "        HT = 2 * p_bar * q_bar\n",
    "        \n",
    "        \n",
    "    elif N == 3:\n",
    "        p_1 = ((2 *pop1_AA ) + pop1_Aa) / (2 * population_1)\n",
    "        q_1= 1 - p_1\n",
    "    \n",
    "        p_2 = ((2* pop2_AA ) + pop2_Aa) / (2 * population_2)\n",
    "        q_2= 1 - p_2\n",
    "        \n",
    "        p_3 = ((2* pop3_AA ) + pop3_Aa) / (2 * population_3)\n",
    "        q_3 = 1 - p_3\n",
    "        \n",
    "        H_obs1 = pop1_Aa / population_1\n",
    "        H_obs2 = pop2_Aa / population_2\n",
    "        H_obs3 = pop3_Aa / population_3\n",
    "        \n",
    "        H_exp_1 = 2 * p_1 * q_1\n",
    "        H_exp_2 = 2 * p_2 * q_2\n",
    "        H_exp_3 = 2 * p_3 * q_3\n",
    "        \n",
    "        F_1 = (H_exp_1 - H_obs1 ) /H_exp_1\n",
    "        F_2 = (H_exp_2 - H_obs2 ) /H_exp_2\n",
    "        F_3 = (H_exp_3 - H_obs3 ) /H_exp_3\n",
    "        \n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3)\n",
    "        p_bar = p_bar_1/  (2 *(population_1 + population_2 + population_3))\n",
    "        \n",
    "        q_bar_1 = (2 *pop1_aa ) + pop1_Aa + (2 *pop2_aa ) + pop2_Aa + (2 *pop3_aa ) + pop3_Aa \n",
    "        q_bar = q_bar_1/ (2 *(population_1 + population_2 + population_3))\n",
    "        \n",
    "        HI_i = (H_obs1 * population_1) + (H_obs2 * population_2) + (H_obs3 * population_3 ) \n",
    "        HI = HI_i / (population_1 + population_2 + population_3)\n",
    "    \n",
    "        HS_s = (H_exp_1 * population_1) + (H_exp_2 * population_2) + (H_exp_3 * population_3 ) \n",
    "        HS = HS_s/ (population_1 + population_2 + population_3)\n",
    "        \n",
    "        HT = 2 * p_bar * q_bar\n",
    "        \n",
    "        \n",
    "    elif N == 4:\n",
    "        p_1 = ((2 *pop1_AA ) + pop1_Aa) / (2 * population_1)\n",
    "        q_1= 1 - p_1\n",
    "    \n",
    "        p_2 = ((2* pop2_AA ) + pop2_Aa) / (2 * population_2)\n",
    "        q_2= 1 - p_2\n",
    "        \n",
    "        p_3 = ((2* pop3_AA ) + pop3_Aa) / (2 * population_3)\n",
    "        q_3 = 1 - p_3\n",
    "        \n",
    "        p_4 = ((2* pop4_AA ) + pop4_Aa) / (2 * population_4)\n",
    "        q_4 = 1 - p_4\n",
    "        \n",
    "        H_obs1 = pop1_Aa / population_1\n",
    "        H_obs2 = pop2_Aa / population_2\n",
    "        H_obs3 = pop3_Aa / population_3\n",
    "        H_obs4 = pop4_Aa / population_4\n",
    "        \n",
    "        H_exp_1 = 2 * p_1 * q_1\n",
    "        H_exp_2 = 2 * p_2 * q_2\n",
    "        H_exp_3 = 2 * p_3 * q_3\n",
    "        H_exp_4 = 2 * p_4 * q_4\n",
    "        \n",
    "        F_1 = (H_exp_1 - H_obs1 ) /H_exp_1\n",
    "        F_2 = (H_exp_2 - H_obs2 ) /H_exp_2\n",
    "        F_3 = (H_exp_3 - H_obs3 ) /H_exp_3\n",
    "        F_4 = (H_exp_4 - H_obs4 ) /H_exp_4\n",
    "        \n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3) + p_4 * (2 * population_4)\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4))\n",
    "        \n",
    "        q_bar_1 = (2 *pop1_aa ) + pop1_Aa + (2 *pop2_aa ) + pop2_Aa + (2 *pop3_aa ) + pop3_Aa +  (2 *pop4_aa ) + pop4_Aa \n",
    "        q_bar = q_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4 ))\n",
    "        \n",
    "        HI_i = (H_obs1 * population_1) + (H_obs2 * population_2) + (H_obs3 * population_3 ) + (H_obs4 * population_4 ) \n",
    "        HI = HI_i / (population_1 + population_2 + population_3 + population_4)\n",
    "    \n",
    "        HS_s = (H_exp_1 * population_1) + (H_exp_2 * population_2) + (H_exp_3 * population_3 ) + (H_exp_4 * population_4 ) \n",
    "        HS = HS_s/ (population_1 + population_2 + population_3 + population_4)\n",
    "        \n",
    "        HT = 2 * p_bar * q_bar\n",
    "        \n",
    "    else:\n",
    "        p_1 = ((2 *pop1_AA ) + pop1_Aa) / (2 * population_1)\n",
    "        q_1= 1 - p_1\n",
    "    \n",
    "        p_2 = ((2* pop2_AA ) + pop2_Aa) / (2 * population_2)\n",
    "        q_2= 1 - p_2\n",
    "        \n",
    "        p_3 = ((2* pop3_AA ) + pop3_Aa) / (2 * population_3)\n",
    "        q_3 = 1 - p_3\n",
    "        \n",
    "        p_4 = ((2* pop4_AA ) + pop4_Aa) / (2 * population_4)\n",
    "        q_4 = 1 - p_4\n",
    "        \n",
    "        p_5 = ((2* pop5_AA ) + pop5_Aa) / (2 * population_5)\n",
    "        q_5 = 1 - p_5\n",
    "        \n",
    "        H_obs1 = pop1_Aa / population_1\n",
    "        H_obs2 = pop2_Aa / population_2\n",
    "        H_obs3 = pop3_Aa / population_3\n",
    "        H_obs4 = pop4_Aa / population_4\n",
    "        H_obs5 = pop5_Aa / population_5\n",
    "        \n",
    "        H_exp_1 = 2 * p_1 * q_1\n",
    "        H_exp_2 = 2 * p_2 * q_2\n",
    "        H_exp_3 = 2 * p_3 * q_3\n",
    "        H_exp_4 = 2 * p_4 * q_4\n",
    "        H_exp_5 = 2 * p_5 * q_5\n",
    "        \n",
    "        F_1 = (H_exp_1 - H_obs1 ) /H_exp_1\n",
    "        F_2 = (H_exp_2 - H_obs2 ) /H_exp_2\n",
    "        F_3 = (H_exp_3 - H_obs3 ) /H_exp_3\n",
    "        F_4 = (H_exp_4 - H_obs4 ) /H_exp_4\n",
    "        F_5 = (H_exp_5 - H_obs5 ) /H_exp_5\n",
    "        \n",
    "        p_bar_1 = p_1 * (2 * population_1) + p_2 * (2 * population_2) + p_3 * (2 * population_3) + p_4 * (2 * population_4)\n",
    "        +(p_5 * (2 * population_5))\n",
    "        p_bar = p_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4 + population_5))\n",
    "        \n",
    "        q_bar_1 = (2 *pop1_aa ) + pop1_Aa + (2 *pop2_aa ) + pop2_Aa + (2 *pop3_aa ) + pop3_Aa +  (2 *pop4_aa ) + pop4_Aa \n",
    "        +(2 *pop5_aa ) + pop5_Aa\n",
    "        q_bar = q_bar_1/ (2 *(population_1 + population_2 + population_3 + population_4 + population_5))\n",
    "        \n",
    "        HI_i = (H_obs1 * population_1) + (H_obs2 * population_2) + (H_obs3 * population_3 ) + (H_obs4 * population_4 ) \n",
    "        + (H_obs5 * population_5 )\n",
    "        HI = HI_i / (population_1 + population_2 + population_3 + population_4 + population_5)\n",
    "    \n",
    "        HS_s = (H_exp_1 * population_1) + (H_exp_2 * population_2) + (H_exp_3 * population_3 ) + (H_exp_4 * population_4 ) \n",
    "        + (H_exp_5 * population_5 )\n",
    "        HS = HS_s/ (population_1 + population_2 + population_3 + population_4 + population_5)\n",
    "        \n",
    "        HT = 2 * p_bar * q_bar\n",
    "        \n",
    "    F_IS = (HS - HI) / HS\n",
    "    F_ST = (HT - HS)/ HT\n",
    "    F_IT = (HT -HI)/ HT\n",
    "    \n",
    "    return F_IS, F_ST, F_IT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = Fst(N = 3, population_1 = 500, population_2 =100, \n",
    "             pop1_AA = 125, pop2_AA = 50,\n",
    "             pop1_Aa = 250, pop2_Aa = 30,\n",
    "             pop1_aa = 125, pop2_aa =20,\n",
    "             population_3 = 1000,\n",
    "             pop3_AA = 100, \n",
    "             pop3_Aa = 500,\n",
    "             pop3_aa = 400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(-0.039307128580946024, 0.03437738731856383, -0.0035784648787744212)\n"
     ]
    }
   ],
   "source": [
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculating Tajima's D\n",
    "# used this:\n",
    "#https://github.com/WhalleyT/tajima/blob/master/tajima.py\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import argparse\n",
    "\n",
    "from math import sqrt\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "def _calculate_pairwise(sequences):\n",
    "    \"\"\"Calculate pi, number of pairwise differences.\"\"\"\n",
    "    for seq in sequences:\n",
    "        if len(seq) != len(sequences[0]):\n",
    "            raise(\"All sequences must have the same length.\")\n",
    "\n",
    "    numseqs = len(sequences)\n",
    "\n",
    "    num = float(numseqs * (numseqs - 1)) / float(2)\n",
    "\n",
    "    combos = combinations(sequences, 2)\n",
    "    counts = []\n",
    "    for pair in combos:\n",
    "        seqA = pair[0]\n",
    "        seqB = pair[1]\n",
    "        count = sum(1 for a, b in zip(seqA, seqB) if a != b)\n",
    "        counts.append(count)\n",
    "\n",
    "    return  (float(sum(counts)) / float(num))\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _calculate_segregating_sites(sequences):\n",
    "    \"\"\"Calculate S, number of segregation sites).\"\"\"\n",
    "    # Assume if we're in here seqs have already been checked\n",
    "    combos = combinations(sequences, 2)\n",
    "    indexes = []\n",
    "    for pair in combos:\n",
    "        seqA = pair[0]\n",
    "        seqB = pair[1]\n",
    "        for idx, (i, j) in enumerate(zip(seqA, seqB)):\n",
    "            if i != j:\n",
    "                indexes.append(idx)\n",
    "\n",
    "    indexes = list(set(indexes))\n",
    "\n",
    "    S = len(indexes)\n",
    "    n = len(sequences)\n",
    "\n",
    "    denom = 0\n",
    "    for i in range(1, n):\n",
    "        denom += (float(1) / float(i))\n",
    "\n",
    "    return float(S / denom)\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "#putting everything together\n",
    "# Test against neutrality\n",
    "\n",
    "def Tajima_D(sequences):\n",
    "    for seq in sequences:\n",
    "        if len(seq) != len(sequences[0]):\n",
    "            raise(\"All sequences must have the same length.\")\n",
    "\n",
    "    numseqs = len(sequences)\n",
    "\n",
    "    num = float(numseqs * (numseqs - 1)) / float(2)\n",
    "\n",
    "    combos = combinations(sequences, 2)\n",
    "    counts = []\n",
    "    for pair in combos:\n",
    "        seqA = pair[0]\n",
    "        seqB = pair[1]\n",
    "        count = sum(1 for a, b in zip(seqA, seqB) if a != b)\n",
    "        counts.append(count)\n",
    "    A  = (float(sum(counts)) / float(num))\n",
    "    \n",
    "    # Assume if we're in here seqs have already been checked\n",
    "    \n",
    "    combos = combinations(sequences, 2)\n",
    "    indexes = []\n",
    "    for pair in combos:\n",
    "        seqA = pair[0]\n",
    "        seqB = pair[1]\n",
    "        for idx, (i, j) in enumerate(zip(seqA, seqB)):\n",
    "            if i != j:\n",
    "                indexes.append(idx)\n",
    "\n",
    "    indexes = list(set(indexes))\n",
    "\n",
    "    S = len(indexes)\n",
    "    n = len(sequences)\n",
    "\n",
    "    denom = 0\n",
    "    for i in range(1, n):\n",
    "        denom += (float(1) / float(i))\n",
    "\n",
    "    B =  float(S / denom)\n",
    "    \n",
    "    D = A -B\n",
    "    return D\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.07999999999999963\n"
     ]
    }
   ],
   "source": [
    "A = Tajima_D(sequences = ('AGTTA AGTTC ATTCT AGAGT', 'AGTTG AGTTG ATTCT AGACT','AGTTA AGTTG ATTCT AGACT', \n",
    "                                    'AGATA AGTTC ATTCT AGACT', 'AGATA AGTTG ATTCT AGACT'))\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "#using example from wikipedia\n",
    "#https://en.wikipedia.org/wiki/Tajima%27s_D\n",
    "\n",
    "A = Tajima_D(sequences =('00000 00000 00000 00000',\n",
    "                       '00100 00000 00100 00010',\n",
    "                       '00000 00000 00100 00010',\n",
    "                       '00000 01000 00000 00010',\n",
    "                       '00000 01000 00100 00010'))\n",
    "print(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# haploid diversity?\n",
    "#homozygosity\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chi squared?\n",
    "#hardy weinberg law ?\n",
    "#may include expected values?\n",
    "\n",
    "\n",
    "\n",
    "#ANOVA/AMOVA\n",
    "#https://www.kaggle.com/hamelg/python-for-data-26-anova?\n",
    "#https://popgen.nescent.org/PopDiffSequenceData.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://colab.research.google.com/drive/1o24ZfqDoBEwvRSx_Br1u7edPPXLLx4w7?usp=sharing#scrollTo=U4yJxCMYW79A\n",
    "\n",
    "#PCA\n",
    "#https://www.youtube.com/watch?v=-PCKK_nwFdA&t=21s\n",
    "#38. the video\n",
    "import numpy as np\n",
    "from sklearn import decomposition\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#data needs to be in a matrix???\n",
    "\n",
    "pca = decompositon.PCA(n_components = 2)\n",
    "pca.fit(matrix)\n",
    "print(pca.explained_variance_ratio)\n",
    "to_plot = pca.transform(matrix)\n",
    "plt.scatter(x = to_plot[:, 0], y = to_plot[:, 1]) \n",
    "\n",
    "#maybe try this at plotly?\n",
    "#maybe altair for nicer graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
